x-pack/plugins/fleet/server/services/agents/update_agent_tags_action_runner.ts

```diff
@@ -57,70 +83,125 @@ export async function updateTagsBatch(
     tagsToRemove: string[];
     actionId?: string;
     total?: number;
+    kuery?: string;
   }
-): Promise<{ actionId: string }> {
+): Promise<{ actionId: string; updated?: number; took?: number }> {
   const errors: Record<Agent['id'], Error> = { ...outgoingErrors };
+  const hostedAgentError = `Cannot modify tags on a hosted agent`;
 
   const filteredAgents = await filterHostedPolicies(
     soClient,
     givenAgents,
     errors,
-    `Cannot modify tags on a hosted agent`
+    hostedAgentError
   );
+  const agentIds = filteredAgents.map((agent) => agent.id);
+
+  let query: estypes.QueryDslQueryContainer | undefined;
+  if (options.kuery !== undefined) {
+    const hostedPolicies = await agentPolicyService.list(soClient, {
+      kuery: `${AGENT_POLICY_SAVED_OBJECT_TYPE}.is_managed:true`,
+    });
+    const hostedIds = hostedPolicies.items.map((item) => item.id);
+
+    query = getElasticsearchQuery(options.kuery, false, false, hostedIds);
+  } else {
+    query = {
+      terms: {
+        _id: agentIds,
+      },
+    };
+  }
 
-  const getNewTags = (agent: Agent): string[] => {
-    const existingTags = agent.tags ?? [];
-
-    if (options.tagsToAdd.length === 1 && options.tagsToRemove.length === 1) {
-      const removableTagIndex = existingTags.indexOf(options.tagsToRemove[0]);
-      if (removableTagIndex > -1) {
-        const newTags = uniq([
-          ...existingTags.slice(0, removableTagIndex),
-          options.tagsToAdd[0],
-          ...existingTags.slice(removableTagIndex + 1),
-        ]);
-        return newTags;
-      }
-    }
-    return uniq(difference(existingTags, options.tagsToRemove).concat(options.tagsToAdd));
-  };
-
-  await bulkUpdateAgents(
-    esClient,
-    filteredAgents.map((agent) => ({
-      agentId: agent.id,
-      data: {
-        tags: getNewTags(agent),
+  let res;
+  try {
+    res = await esClient.updateByQuery({
+      query,
+      index: AGENTS_INDEX,
+      refresh: true,
+      wait_for_completion: true,
+      script: {
+        source: `   
+      ctx._source.tags = ctx._source.tags == null ? [] : ctx._source.tags;
+      if (params.tagsToAdd.length == 1 && params.tagsToRemove.length == 1) { 
+        ctx._source.tags.replaceAll(tag -> params.tagsToRemove[0] == tag ? params.tagsToAdd[0] : tag);
+      } else {
+        ctx._source.tags.removeAll(params.tagsToRemove); 
+        ctx._source.tags.addAll(params.tagsToAdd);
```

Comment:
Why no just having that code block, instead of the replaceAll before?

Reply:
In the use case of renaming a tag, we don't want the order to change. This is covered by the `replaceAll` logic to replace the tag in place.